src/Confluent.Kafka/IProducer.cs

```diff
@@ -0,0 +1,149 @@
+// limitations under the License.
+//
+// Refer to LICENSE for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Linq;
+using System.Runtime.InteropServices;
+using Confluent.Kafka.Impl;
+using Confluent.Kafka.Internal;
+using Confluent.Kafka.Serialization;
+using System.Collections.Concurrent;
+
+
+namespace Confluent.Kafka
+{
+    /// <summary>
+    ///     Defines a high-level Apache Kafka producer client (without 
+    ///     serialization capability).
+    /// </summary>
+    public interface IProducer : IDisposable
+    {
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll_int"]/*' />
+        int Poll(int millisecondsTimeout);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll_TimeSpan"]/*' />
+        int Poll(TimeSpan timeout);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="Poll"]/*' />
+        [Obsolete("Use an overload of Poll with a finite timeout.", false)]
+        int Poll();
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="GetSerializingProducer"]/*' />
+        ISerializingProducer<TKey, TValue> GetSerializingProducer<TKey, TValue>(ISerializer<TKey> keySerializer, ISerializer<TValue> valueSerializer);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Message"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(Message message);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_TKey_TValue"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(string topic, byte[] key, byte[] val);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_TKey_TValue_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(
+            string topic, Partition partition, 
+            byte[] key, byte[] val, 
+            Timestamp timestamp, 
+            IEnumerable<KeyValuePair<string, byte[]>> headers);
+
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_string_Partition_byte_int_int_byte_int_int_Timestamp_IEnumerable"]/*' />
+        /// <include file='include_docs_producer.xml' path='API/Member[@name="ProduceAsync_Common"]/*' />
+        Task<Message> ProduceAsync(
+            string topic, Partition partition, 
+            byte[] key, int keyOffset, int keyLength, 
+            byte[] val, int valOffset, int valLength, 
+            Timestamp timestamp, IEnumerable<KeyValuePair<string, byte[]>> headers);
```

Comment:
In the non-serializing producer case I'm now thinking we should consider removing all `Produce` variants except this one (i.e. the `Message` one as well`). I don't expect people to use this class, except really for the `GetSerializingProducer` method, or for super optimization avoiding memory allocations (which this method allows them to do).

In the serializing producer case, I'm now thinking we should actually get rid of the `Message` variant. Accepting a `Message` class in the produce method really just shifts the problem of lots of producer method overloads to lots of `Message` class constructor overloads. Admittedly, it's half the problem there (since there are no interfaces backing `Message`, so there is no duplication due to that), but the UX is far worse having to construct a `Message` class in order to produce.

There is a use case for people constructing `Messages` in one place, then sending the message in another, which the `Produce` `Message` overload is useful for, but i suspect that is not that common and it's not too hard to code that functionality even if we don't supply the `Message` variant of `Produce`.

Final comment here: I wish `Message` was called `DeliveryReport` (but I don't suggest changing it). It combines the what, with the where, with the error field. `Message` to me implies just the 'what'. It would be better if these were separated more explicitly imo. By avoiding using the `Message` class in the `Produce` calls, we avoid some Matt anguish due to this.

Reply:
>  I'm now thinking we should actually get rid of the Message variant. 

The problem is when a new message field is added (which seems to happen 1-2 times a year), then there need to be yet another non-Message Produce() method.
Having the Message evolve, rather than the Produce call, makes more sense to me since that is actually what is changing underneath: message attributes, not produce semantics.

Reply:
Note that I'm suggesting just **two** variants of produce/produceAsync on the serializing producer (4 ways of producing in total).

- The first just takes topic, key and value (basic usage, often used).
- The second takes everything. This is actually pretty convenient to use even if you want to specify a subset of the parameters, because I can write, eg `Partition.Any`, `Timestamp.Default`, and it's all very readable and quite short.

I think we likely won't need any more than this number of methods. we can add to the "includes all" variant if future parameters are required (and specify default values for backwards compatibility). 

are you aware of anything coming up?

On the non-serializing producer, I'm suggesting just a single variant (of produce/produceAsync) that takes all parameters.

I get where you're coming from, but I don't think it's the best thing to be focussed on.